### PR TITLE
Add COLLABORATOR_MODE event.

### PR DESCRIPTION
**Pull request by alstojanovic**
### Describe the background of your pull request

What does your pull request do? Does it solve a bug (which one?), add a feature?

### Additional context

Add any other context

### Governance 

- [ ] Documentation is added
- [ ] Test cases are added or updated

By submitting this pull request I confirm that:
- I've read the contributing document https://github.com/alkem-io/.github/blob/master/CONTRIBUTING.md.
- I've read and understand the Code of Conduct https://github.com/alkem-io/.github/blob/master/CODE_OF_CONDUCT.md.
- I understand that any contributions or suggestions I made may make it into the actual code. I've read the License 
     https://github.com/alkem-io/Coordination/blob/master/LICENSE.
- I understand that submitting the pull request requires agreeing to and signing the contributor license agreement.

### CODE REVIEW

File: src/services/external/excalidraw-backend/excalidraw.server.ts [hunk 1]
```diff
@@ -195,7 +209,12 @@ export class ExcalidrawServer {
         `User '${socket.data.agentInfo.userID}' established connection`,
         LogContext.EXCALIDRAW_SERVER
       );
+
       this.wsServer.to(socket.id).emit(INIT_ROOM);
+      this.wsServer.to(socket.id).emit(COLLABORATOR_MODE, { mode: 'write' });
```
Comment (hero101) [inline src/services/external/excalidraw-backend/excalidraw.server.ts]: the mode is not determined by this point. You are assuming it's always `write`.
emit it where the user is authorized with the room

File: src/services/external/excalidraw-backend/excalidraw.server.ts [hunk 2]
```diff
@@ -238,22 +258,33 @@ export class ExcalidrawServer {
...
           await disconnectingEventHandler(this.wsServer, socket, this.logger)
       );
-      socket.on(DISCONNECT, () => disconnectEventHandler(socket));
+      socket.on(DISCONNECT, () => {
+        disconnectEventHandler(socket);
+        this.deleteCollaboratorModeTimerForSocket(socket.id);
+      });
+    });
+
+    this.whiteboardService.EventEmitter.on('saved', async (roomID: string) => {
```
Comment (hero101) [inline src/services/external/excalidraw-backend/excalidraw.server.ts]: use the constant for the event name.
the function is not async

File: src/services/external/excalidraw-backend/excalidraw.server.ts [hunk 3]
```diff
@@ -408,6 +439,66 @@ export class ExcalidrawServer {
...
+        this.logger.verbose?.(
+          `Reseting collaborator mode timer for socket '${socket.id}'`,
+          LogContext.EXCALIDRAW_SERVER
+        );
+
+        this.deleteCollaboratorModeTimerForSocket(socket.id);
+        this.startCollaboratorModeTimer(socket);
+      }
+    },
+    500,
```
Comment (hero101) [inline src/services/external/excalidraw-backend/excalidraw.server.ts]: magic number; also try with 1 second

File: src/services/external/excalidraw-backend/types/socket.data.ts [hunk 4]
```diff
@@ -11,6 +11,11 @@ export type SocketData = {
    * The events are the ones received on the server-broadcast channel
    */
   lastContributed: number;
+  /***
+   * The timestamp of the last contribution event received;
+   * The events are the ones received on the server-broadcast channel
+   */
+  lastCursorMove: number;
```
Comment (hero101) [inline src/services/external/excalidraw-backend/types/socket.data.ts]: Change the variable to `lastPresence`.
Change the description to
`The timestamp of the last presence received;
These events are received with the 'server-volatile-broadcast' event`

File: src/domain/common/whiteboard/whiteboard.service.ts [hunk 5]
```diff
@@ -29,9 +30,11 @@ import { IWhiteboard } from './whiteboard.interface';
 import { CreateWhiteboardInput } from './dto/whiteboard.dto.create';
 import { UpdateWhiteboardInput } from './dto/whiteboard.dto.update';
 import { WhiteboardAuthorizationService } from './whiteboard.service.authorization';
+import { SAVED } from '@services/external/excalidraw-backend/types/event.names';
 
 @Injectable()
 export class WhiteboardService {
+  public EventEmitter = new EventEmitter();
```
Comment (hero101) [inline src/domain/common/whiteboard/whiteboard.service.ts]: change the variable name

File: src/services/external/excalidraw-backend/excalidraw.server.ts [hunk 6]
```diff
@@ -408,6 +439,66 @@ export class ExcalidrawServer {
...
+      socket.removeAllListeners(SERVER_BROADCAST);
+      socket.data.update = false;
+      this.collaboratorModeTimers.delete(socket.id);
+    }, this.collaboratorModeTimeoutMs);
+  }
+
+  private startCollaboratorModeTimer(socket: SocketIoSocket) {
+    let timer = this.collaboratorModeTimers.get(socket.id);
+    if (timer) {
+      this.logger.verbose?.(
```
Comment (hero101) [inline src/services/external/excalidraw-backend/excalidraw.server.ts]: i am not confinced this logging is needed

File: src/services/external/excalidraw-backend/excalidraw.server.ts [hunk 7]
```diff
@@ -408,6 +439,66 @@ export class ExcalidrawServer {
...
+    }
+    timer = this.createCollaboratorModeTimer(socket);
+    this.collaboratorModeTimers.set(socket.id, timer);
+    this.logger.verbose?.(
+      `Created collaborator mode timer for socket '${socket.id}'`,
+      LogContext.EXCALIDRAW_SERVER
+    );
+  }
+
+  private resetCollaboratorModeTimer = debounce(
```
Comment (hero101) [inline src/services/external/excalidraw-backend/excalidraw.server.ts:474]: try using [timeout.refresh()](https://nodejs.org/docs/latest-v18.x/api/timers.html#timeoutrefresh)

**Review by hero101 — CHANGES_REQUESTED** [review]

File: src/domain/common/whiteboard/whiteboard.service.ts (hunk 5 above)
Comment (alstojanovic) [reply src/domain/common/whiteboard/whiteboard.service.ts]: Any suggestions? This is a generic eventEmmiter for whiteboard service used to send any events.

File: src/services/external/excalidraw-backend/excalidraw.server.ts [hunk 8]
```diff
@@ -195,7 +211,11 @@ export class ExcalidrawServer {
         `User '${socket.data.agentInfo.userID}' established connection`,
         LogContext.EXCALIDRAW_SERVER
       );
+
       this.wsServer.to(socket.id).emit(INIT_ROOM);
+
+      this.startCollaboratorModeTimer(socket);
```
Comment (hero101) [inline src/services/external/excalidraw-backend/excalidraw.server.ts]: start the timer after the user has been authorized.
delete the timer also on socket error - there is a handler about it

File: src/domain/common/whiteboard/whiteboard.service.ts (hunk 5 above)
Comment (hero101) [reply src/domain/common/whiteboard/whiteboard.service.ts]: add some comments to it to explain what is its use - it's a new concept for the backend.
about the name - at least start with small caps
if it generic, emit a well named event, like `content-changed` instead an excalidraw event. They shouldn't have anything in common.

File: src/services/external/excalidraw-backend/excalidraw.server.ts [hunk 9]
```diff
@@ -408,6 +440,61 @@ export class ExcalidrawServer {
...
+      `Created collaborator mode timer for socket '${socket.id}'`,
+      LogContext.EXCALIDRAW_SERVER
+    );
+  }
+
+  private resetCollaboratorModeTimer = debounce(
+    (socket: SocketIoSocket) => {
+      const timer = this.collaboratorModeTimers.get(socket.id);
+      if (timer) {
+        this.logger.verbose?.(
```
Comment (hero101) [inline src/services/external/excalidraw-backend/excalidraw.server.ts]: that would be too much logging, change it to debug or discard it

**Review by hero101 — CHANGES_REQUESTED** [review]

**Review by hero101 — APPROVED** [review]

**Pull request merged**